### PR TITLE
Update Writing Documentation.md

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

**Comment by josevalim** [comment]
We now talk about documentation metadata and I have unified the discussion about private docs and code comments.

File: lib/elixir/pages/Writing Documentation.md [hunk 1]
```diff
@@ -51,6 +51,28 @@ The compiler will infer this argument as `map`. Sometimes the inference will be
       size
     end
 
+## Documentation metadata
+
+Elixir allows developers to attach random metadata to the documentation too. This is done by using the relevant attribute, such as `@moduledoc`, `@typedoc` and `@doc` and passing a keyword list where the key is the metadata key and value is the metadata value. A commonly used metadata is `:since`, which annotates which version that particular module, function, type or callback was added. For instance, in the example above we could have added:
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Elixir allows developers to attach random metadata to the documentation too. This is done by using the relevant attribute (such as `@moduledoc`, `@typedoc`, and `@doc`) and passing a keyword list where the key is the metadata key and the value is the metadata value. A commonly used metadata is `:since`, which annotates in which version that particular module, function, type, or callback was added. For instance, in the example above we could have added:
```

File: lib/elixir/pages/Writing Documentation.md [hunk 2]
```diff
@@ -51,6 +51,28 @@ The compiler will infer this argument as `map`. Sometimes the inference will be
...
+
+    defmodule MyApp.Hello do
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]: What do you think about giving a more complete example?

```elixir
defmodule MyApp.HelloWorld do
  @moduledoc "Says hello to the world"
  @moduledoc since: "1.0.0"

  @doc "Prints hello world"
  @doc since: "1.3.0"
  def hello_world, do: IO.puts("Hello world!")
end
```

File: lib/elixir/pages/Writing Documentation.md [hunk 3]
```diff
@@ -51,6 +51,28 @@ The compiler will infer this argument as `map`. Sometimes the inference will be
...
+
+Another existing metadata is `:deprecated`, which emits a warning in the documentation that usage of the function is discouraged:
+
+    @doc deprecated: "Use Foo.bar/2 instead"
+
+Note the `:deprecated` does not warn when a developer invokes the functions. If you want the code to also emit a warning, you can use the `@deprecated` attribute:
+
+    @deprecated "Use Foo.bar/2 instead"
+
+Metadata can have any key and is often by documentation tools to provide more accessible ways to traverse your project's documentation.
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Metadata can have any key and is often used by documentation tools to provide more accessible ways to traverse your project's documentation.
```

And maybe the first part can be moved to the top, to something like:

> Elixir allows developers to attach random metadata to the documentation too. This is done by using the relevant attribute (such as `@moduledoc`, `@typedoc`, and `@doc`) and passing a keyword list where the key is the metadata key and the value is the metadata value:
>
> ```elixir
> @doc "Returns the meaning of life"
> @doc since: "1.0.1", authors: ["John Doe"]
> def meaning_of_life, do: 42
> ```
>
> A commonly used metadata is `:since`, which annotates in which version that particular module, function, type, or callback was added. For instance, in the example above we could have added:
>
> ...

File: lib/elixir/pages/Writing Documentation.md [hunk 4]
```diff
@@ -70,26 +92,35 @@ When writing documentation:
...
 Elixir treats documentation and code comments as different concepts. Documentation is for users of your Application Programming Interface (API), be it your co-worker or your future self. Modules and functions must always be documented if they are part of your API.
 
 Code comments are for developers reading the code. They are useful to mark improvements, leave notes for developers reading the code (for example, you decided not to call a function due to a bug in a library) and so forth.
 
-In other words: documentation is required, code comments are optional.
+Because private functions cannot be accessed externally and effectively exist only in your source code, you can't attach a `@doc` attribute to private functions. However, you can add code comments to private functions, as with any other piece of code, and we recommend developers to do so whenever they believe it will add relevant information to readers and maintainers of the code.
+
+Finally, beware of redundant code comments, such as code comments that are describing the exact same as the code:
+
+    # Total is the sum of the batch and invidual entries
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
    # Total is the sum of the batch and individual entries
```

File: lib/elixir/pages/Writing Documentation.md [hunk 5]
```diff
@@ -70,26 +92,35 @@ When writing documentation:
...
+    total = batch_sum + individual_sum
+
+In summary, documentation are for users of your API (who do not necessarily have access to source) and code comments are for those who interact with the source.
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]: Not 100% sure about this one, but this sounds better to me

```suggestion
In summary, documentation are for users of your API (who do not necessarily have access to the source code) and code comments are for those who interact with the source code.
```

File: lib/elixir/pages/Writing Documentation.md [hunk 6]
```diff
@@ -70,26 +92,35 @@ When writing documentation:
...
 
 ## Hiding Internal Modules and Functions
 
 Besides the modules and functions libraries provide as part of their public interface, libraries may also implement important functionality that is not part of their API. While these modules and functions can be accessed, they are meant to be internal to the library and thus should not have documentation for end users.
 
-Luckily, Elixir allows developers to hide modules and functions from the documentation. For example, one common practice for documenting internal behaviour is to set the `@moduledoc` attribute to `false` while documenting each function:
+Luckily, Elixir allows developers to hide modules and functions from the documentation by setting `@doc false`, to hide a particular function, or `@moduledoc false` to hide the whole module. If a module is hidden, you may even document the functions in the module, but the module itself nor any of its functions will be included in the documentation:
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Luckily, Elixir allows developers to hide modules and functions from the documentation by setting `@doc false` to hide a particular function, or `@moduledoc false` to hide the whole module. If a module is hidden, you may even document the functions in the module, but the module itself nor any of its functions will be included in the documentation:
```

File: lib/elixir/pages/Writing Documentation.md [hunk 7]
```diff
@@ -102,25 +133,12 @@ Luckily, Elixir allows developers to hide modules and functions from the documen
...
-
 However, keep in mind that adding `@doc false` does not make the function private. The function above can still be invoked as `MyApp.Sample.add(1, 2)`. Not only that, if `MyApp.Sample` is imported, the `add/2` function will also be imported into the caller. For those reasons, be cautious when adding `@doc false` to functions, instead use one of these two options:
 
   * Move the undocumented function to a module with `@moduledoc false`, like `MyApp.Hidden`, ensuring the function won't be accidentally exposed or imported. Remember you can use `@moduledoc false` to hide a whole module and still document each function with `@doc`. Tools will still ignore the module.
 
   * Start the function name with one or two underscores, for example, `__add__/2`, and add `@doc false`. The compiler does not import functions with leading underscores and they hint to anyone reading the code of their intended private usage.
 
-## Documenting Private Functions
-
-Elixir warns if a private function has a `@doc` attribute and discards its content, because `@doc` is intended to be used only for your public interface.
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]: The info:

> Elixir warns if a private function has a `@doc` attribute and discards its content

is missed in the "Code comments" section and I think it's valuable 🙂

File: lib/elixir/pages/Writing Documentation.md [hunk 8]
```diff
@@ -70,26 +92,35 @@ When writing documentation:
   * Start new sections with second level Markdown headers `##`. First level headers are reserved for module and function names.
 
   * Place documentation before the first clause of multi-clause functions. Documentation is always per function and arity and not per clause.
+  
+  * Use the `:since` key in the documentation metadata to annotate whenever new functions are added to your API.
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
  * Use the `:since` key in the documentation metadata to annotate whenever new functions or modules are added to your API.
```

?

File: lib/elixir/pages/Writing Documentation.md (hunk 5 above)
Comment (eksperimental) [reply lib/elixir/pages/Writing Documentation.md]: I dont like the use of "might", because I find irrelevant whether you have access to the source or not. If you are reading the documentation, why would you read the source code unless you have a very good reason for it.

Comment (eksperimental) [reply lib/elixir/pages/Writing Documentation.md]: One more correction:
In summary, documentation **is**

File: lib/elixir/pages/Writing Documentation.md (hunk 2 above)
Comment (josevalim) [reply lib/elixir/pages/Writing Documentation.md]: There is an example at the top of the file that I am referring to but you can't see it in a diff. Isn't that enough?

File: lib/elixir/pages/Writing Documentation.md (hunk 7 above)
Comment (josevalim) [reply lib/elixir/pages/Writing Documentation.md]: Done, thanks!

File: lib/elixir/pages/Writing Documentation.md [hunk 9]
```diff
@@ -17,6 +17,7 @@ Documentation in Elixir is usually attached to module attributes. Let's see an e
       @moduledoc """
       This is the Hello module.
       """
+      @moduledoc since: "1.0.0"
```
Comment (josevalim) [inline lib/elixir/pages/Writing Documentation.md]: @fertapric i moved it up for completeness. wdty?

**Comment by josevalim** [comment]
Thanks for the great review @fertapric and @eksperimental !

File: lib/elixir/pages/Writing Documentation.md (hunk 9 above)
Comment (fertapric) [reply lib/elixir/pages/Writing Documentation.md]: You beat me by seconds, I was about to suggest the same! 👍

File: lib/elixir/pages/Writing Documentation.md [hunk 10]
```diff
@@ -51,6 +53,20 @@ The compiler will infer this argument as `map`. Sometimes the inference will be
       size
     end
 
+## Documentation metadata
+
+Elixir allows developers to attach random metadata to the documentation. This is done by passing a keyword list to the relevant attribute (such as `@moduledoc`, `@typedoc`, and `@doc`). A commonly used metadata is `:since`, which annotates in which version that particular module, function, type, or callback was added on, as shown in the example above.
```
Comment (ericmj) [inline lib/elixir/pages/Writing Documentation.md]: Change "random metadata" to "any metadata" or "arbitrary metadata"?

Comment (ericmj) [inline lib/elixir/pages/Writing Documentation.md]: or callback was added ~~on~~

File: lib/elixir/pages/Writing Documentation.md [hunk 11]
```diff
@@ -70,26 +86,35 @@ When writing documentation:
...
+Code comments are for developers reading the code. They are useful to mark improvements, leave notes for developers reading the code (for example, why you had to resort to a workaround due to a bug in a library) and so forth. Code comments are tied to the source code. You can completely rewrite a function, remove all existing code comments, and it can continue to behave the same, with no change to its behaviour or its documentation.
 
-Code comments are for developers reading the code. They are useful to mark improvements, leave notes for developers reading the code (for example, you decided not to call a function due to a bug in a library) and so forth.
+Because private functions cannot be accessed externally and effectively exist only in your source code, Elixir will warns if a private function has a `@doc` attribute and discards its content. However, you can add code comments to private functions, as with any other piece of code, and we recommend developers to do so whenever they believe it will add relevant information to readers and maintainers of the code.
```
Comment (ericmj) [inline lib/elixir/pages/Writing Documentation.md]: Elixir will warn~~s~~

File: lib/elixir/pages/Writing Documentation.md [hunk 12]
```diff
@@ -70,26 +86,35 @@ When writing documentation:
...
 ## Doctests
 
 We recommend that developers include examples in their documentation, often under their own `## Examples` heading. To ensure examples do not get out of date, Elixir's test framework (ExUnit) provides a feature called doctests that allows developers to test the examples in their documentation. Doctests work by parsing out code samples starting with `iex>` from the documentation. You can read more about it at `ExUnit.DocTest`.
 
 Notice doctests have limitations. When you cannot doctest a function, because it relies on state or side-effects, we recommend developers include examples directly without the `iex>` prompt.
 
-## Documentation != Comments
+## Documentation != Code comments
+
+Elixir treats documentation and code comments as different concepts. Documentation is a explicit contract between you and users of your Application Programming Interface (API), be them 3rd-party developers, co-workers or your future self. Modules and functions must always be documented if they are part of your API.
```
Comment (eksperimental) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Elixir treats documentation and code comments as different concepts. Documentation is a explicit contract between you and users of your Application Programming Interface (API), be them third-party developers, co-workers or your future self. Modules and functions must always be documented if they are part of your API.
```

We have replaced all these cases with plain letters

File: lib/elixir/pages/Writing Documentation.md [hunk 13]
```diff
@@ -70,26 +86,35 @@ When writing documentation:
...
 
-Elixir treats documentation and code comments as different concepts. Documentation is for users of your Application Programming Interface (API), be it your co-worker or your future self. Modules and functions must always be documented if they are part of your API.
+Code comments are for developers reading the code. They are useful to mark improvements, leave notes for developers reading the code (for example, why you had to resort to a workaround due to a bug in a library) and so forth. Code comments are tied to the source code. You can completely rewrite a function, remove all existing code comments, and it can continue to behave the same, with no change to its behaviour or its documentation.
```
Comment (eksperimental) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Code comments are aimed at developers reading the code. They are useful to mark improvements, leave notes (for example, why you had to resort to a workaround due to a bug in a library), and so forth. Code comments are tied to the source code. You can completely rewrite a function, remove all existing code comments, and it will continue to behave the same, with no change to either its behaviour or its documentation.
```

File: lib/elixir/pages/Writing Documentation.md (hunk 12 above)
Comment (eksperimental) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Elixir treats documentation and code comments as different concepts. Documentation is a explicit contract between you and users of your Application Programming Interface (API), be them 3rd-party developers, co-workers or your future self. Modules and functions should always be documented if they are part of your API.
```

File: lib/elixir/pages/Writing Documentation.md [hunk 14]
```diff
@@ -70,26 +86,35 @@ When writing documentation:
...
 
 ## Hiding Internal Modules and Functions
 
 Besides the modules and functions libraries provide as part of their public interface, libraries may also implement important functionality that is not part of their API. While these modules and functions can be accessed, they are meant to be internal to the library and thus should not have documentation for end users.
 
-Luckily, Elixir allows developers to hide modules and functions from the documentation. For example, one common practice for documenting internal behaviour is to set the `@moduledoc` attribute to `false` while documenting each function:
+Luckily, Elixir allows developers to hide modules and functions from the documentation, by setting `@doc false` to hide a particular function, or `@moduledoc false` to hide the whole module. If a module is hidden, you may even document the functions in the module, but neither the module itself nor any of its functions will be included in the documentation:
```
Comment (eksperimental) [inline lib/elixir/pages/Writing Documentation.md]: > by setting `@doc false` to hide a particular function, or `@moduledoc false` to hide the whole module. If a module is hidden, you may even document the functions in the module, but neither the module itself nor any of its functions will be included in the documentation:

This is not what actually happens in IEx, where a module with `@moduledoc false` and functions documented with `@doc` can be accessed. Try in IEx `h IEx.History.<TAB>` to see what I mean.

File: lib/elixir/pages/Writing Documentation.md (hunk 10 above)
Comment (josevalim) [reply lib/elixir/pages/Writing Documentation.md]: So @eksperimental suggested to add "on" and since I don't know which way is correct, I asked Chris and he said without "on" is correct. So I will remove it.

File: lib/elixir/pages/Writing Documentation.md (hunk 12 above)
Comment (josevalim) [reply lib/elixir/pages/Writing Documentation.md]: I believe "must" is better here because it gives a strong suggestion. Are you concerned that the must implies the compiler will error otherwise or something of sorts?

File: lib/elixir/pages/Writing Documentation.md [hunk 15]
```diff
@@ -51,6 +53,20 @@ The compiler will infer this argument as `map`. Sometimes the inference will be
       size
     end
 
+## Documentation metadata
+
+Elixir allows developers to attach arbitrary metadata to the documentation. This is done by passing a keyword list to the relevant attribute (such as `@moduledoc`, `@typedoc`, and `@doc`). A commonly used metadata is `:since`, which annotates in which version that particular module, function, type, or callback was added, as shown in the example above.
+
+Another common metadata is `:deprecated`, which emits a warning in the documentation, explaining that its usage of is discouraged:
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Another common metadata is `:deprecated`, which emits a warning in the documentation, explaining that its usage is discouraged:
```

File: lib/elixir/pages/Writing Documentation.md [hunk 16]
```diff
@@ -70,26 +86,35 @@ When writing documentation:
...
 ## Doctests
 
 We recommend that developers include examples in their documentation, often under their own `## Examples` heading. To ensure examples do not get out of date, Elixir's test framework (ExUnit) provides a feature called doctests that allows developers to test the examples in their documentation. Doctests work by parsing out code samples starting with `iex>` from the documentation. You can read more about it at `ExUnit.DocTest`.
 
 Notice doctests have limitations. When you cannot doctest a function, because it relies on state or side-effects, we recommend developers include examples directly without the `iex>` prompt.
 
-## Documentation != Comments
+## Documentation != Code comments
+
+Elixir treats documentation and code comments as different concepts. Documentation is a explicit contract between you and users of your Application Programming Interface (API), be them third-party developers, co-workers or your future self. Modules and functions must always be documented if they are part of your API.
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Elixir treats documentation and code comments as different concepts. Documentation is an explicit contract between you and users of your Application Programming Interface (API), be them third-party developers, co-workers, or your future self. Modules and functions must always be documented if they are part of your API.
```

File: lib/elixir/pages/Writing Documentation.md [hunk 17]
```diff
@@ -70,26 +86,35 @@ When writing documentation:
...
+Code comments are aimed at developers reading the code. They are useful to mark improvements, leave notes (for example, why you had to resort to a workaround due to a bug in a library), and so forth. Code comments are tied to the source code. You can completely rewrite a function, remove all existing code comments, and it will continue to behave the same, with no change to either its behaviour or its documentation.
 
-Code comments are for developers reading the code. They are useful to mark improvements, leave notes for developers reading the code (for example, you decided not to call a function due to a bug in a library) and so forth.
+Because private functions cannot be accessed externally and effectively exist only in your source code, Elixir will warn if a private function has a `@doc` attribute and will discard its content. However, you can add code comments to private functions, as with any other piece of code, and we recommend developers to do so whenever they believe it will add relevant information to the readers and maintainers of such code.
```
Comment (fertapric) [inline lib/elixir/pages/Writing Documentation.md]:
```suggestion
Because private functions cannot be accessed externally, Elixir will warn if a private function has a `@doc` attribute and will discard its content. However, you can add code comments to private functions, as with any other piece of code, and we recommend developers to do so whenever they believe it will add relevant information to the readers and maintainers of such code.
```

Or I would rephrase it because it might be read as "Because private functions cannot ... effectively exist only in your source code"

File: lib/elixir/pages/Writing Documentation.md (hunk 12 above)
Comment (eksperimental) [reply lib/elixir/pages/Writing Documentation.md]: Exactly, "must" gives the impression that I wouldn't compile.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
